zookeeper-server/src/main/java/org/apache/zookeeper/backup/NFSBackupStorage.java

```
@@ -0,0 +1,324 @@
+  private static final String PATH_SEPARATOR = File.separator;
+  private final BackupConfig _backupConfig;
+  private final String _fileRootPath;
+
+  /**
+   * Constructor using BackupConfig to get backup storage info
+   * @param backupConfig The information and settings about backup storage, to be set as a part of ZooKeeper server config
+   */
+  public NFSBackupStorage(BackupConfig backupConfig) {
+    _backupConfig = backupConfig;
+    _fileRootPath =
+        String.join(PATH_SEPARATOR, _backupConfig.getMountPath(), _backupConfig.getNamespace());
+  }
+
+  @Override
+  public BackupFileInfo getBackupFileInfo(File file) throws IOException {
+    String backupFilePath = constructBackupFilePath(file.getName());
+    File backupFile = new File(backupFilePath);
+
+    if (!backupFile.exists()) {
+      throw new BackupException(
+          "Backup file " + file.getName() + " does not exist, could not get file info.");
+    }
+
+    BasicFileAttributes fileAttributes =
+        Files.readAttributes(Paths.get(backupFilePath), BasicFileAttributes.class);
+    return new BackupFileInfo(backupFile, fileAttributes.lastModifiedTime().toMillis(),
+        fileAttributes.size());
+  }
+
+  @Override
+  public List<BackupFileInfo> getBackupFileInfos(File path, String prefix) throws IOException {
+    String backupDirPath = constructBackupFilePath(path.getPath());
+    File backupDir = new File(backupDirPath);
+
+    if (!backupDir.exists()) {
+      throw new BackupException(
+          "Backup directory " + path.getPath() + " does not exist, could not get file info.");
+    }
+
+    // Get a list of files whose file names start with the provided prefix
+    FilenameFilter fileFilter = (dir, name) -> name.startsWith(prefix);
+    File[] files = backupDir.listFiles(fileFilter);
+    if (files == null) {
+      throw new BackupException("The provided directory path " + path.getPath()
+          + " is invalid, could not get file info.");
+    }
```

Why is it necessary that we throw an exception here instead of return an empty list?

---

If the files is null, meaning the path is not valid. If it's just empty directory, it will return an empty array instead of null, and in that case, we will just return an empty list.
I just thought it might be worth it to bring to our attention that the path we try to access is invalid, so I throw the exception here. Please let me know if you think it's unnecessary.

---

I thought returning an exception in that case might be an overkill - might require the caller to catch more exceptions. But if the path is not valid, it might be worth throwing an exception.

When would this path ever be not valid?

---

The path is invalid means it is not we could not locate a file / directory using that path.